### PR TITLE
[project-base] optimization of ProductDataFixture.php

### PR DESCRIPTION
**Pull request by RostislavKreisinger**
| Q             | A
| ------------- | ---
|Description, reason for the PR| speed-up product demo data loading approximately to one quarter of the original time , little optimize caching product entities for building product variants - now cache only necessary products instead of all.
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: project-base/src/DataFixtures/Demo/ProductDataFixture.php [hunk 1]
```diff
@@ -5836,4 +5857,20 @@ public function getDependencies(): array
             SettingValueDataFixture::class,
         ];
     }
+
+    /**
+     * @param array $array
+     * @return string[]
+     */
+    private function arrayFlat(array $array): array
```
Comment (TomasLudvik) [inline project-base/src/DataFixtures/Demo/ProductDataFixture.php:5865]:
```suggestion
    private function getAllVariantCatnumsFromAssociativeArray(array $productCatnumsByMainVariantCatnum): array
```

File: project-base/src/DataFixtures/Demo/ProductDataFixture.php [hunk 2]
```diff
@@ -5836,4 +5857,20 @@ public function getDependencies(): array
...
+    {
+        $result = [];
```
Comment (TomasLudvik) [inline project-base/src/DataFixtures/Demo/ProductDataFixture.php:5867]:
```suggestion
        $catnums = [];
```

File: project-base/src/DataFixtures/Demo/ProductDataFixture.php [hunk 3]
```diff
@@ -5836,4 +5857,20 @@ public function getDependencies(): array
...
+
+        foreach ($array as $key => $values) {
```
Comment (TomasLudvik) [inline project-base/src/DataFixtures/Demo/ProductDataFixture.php:5869]:
```suggestion
        foreach ($productCatnumsByMainVariantCatnum as $mainVariantCatnum => $variantCatnums) {
```

File: project-base/src/DataFixtures/Demo/ProductDataFixture.php [hunk 4]
```diff
@@ -72,26 +73,26 @@ class ProductDataFixture extends AbstractReferenceFixture implements DependentFi
      */
     private $parameterDataFactory;
 
-    /**
-     * @var \Shopsys\FrameworkBundle\Model\Product\Parameter\Parameter[]
-     */
-    private $parameters;
-
```
Comment (TomasLudvik) [inline project-base/src/DataFixtures/Demo/ProductDataFixture.php:79]: Why have been parameters removed here?

**Comment by TomasLudvik** [comment]
I would also moved this to version 9.0

**Comment by TomasLudvik** [comment]
I have rebased it on version 9, so it is testable and tests can be compared.

I have been running `php phing db-wipe-public-schema db-import-basic-structure db-migrations domains-data-create` before every test, so I have fresh database prepared for data fixtures and then `php phing db-fixtures-demo` to install and measure data fixtures.

Before change:
first run - 35.6149 seconds
second run - 36.5321 seconds
third run - 36.6018 seconds

After change:
first run - 34.8728 seconds
second run - 34.6371 seconds
third run - 34.6887 seconds

We can say, that this change speed it up ~ 1 - 2 seconds.

**Comment by TomasLudvik** [comment]
Oh and UPGRADE notes are missing

File: project-base/src/DataFixtures/Demo/ProductDataFixture.php (hunk 4 above)
Comment (s3tezsky) [reply project-base/src/DataFixtures/Demo/ProductDataFixture.php:79]: I guess the main reason is due to persistence reference.

**Pull request merged**